qutebrowser/misc/sql.py

```diff
@@ -475,3 +476,87 @@ def select(self, sort_by, sort_order, limit=-1):
                           sort_order=sort_order))
         q.run(limit=limit)
         return q
+
+
+class Database:
+
+    _database: Optional[QSqlDatabase] = None
+    _user_version: Optional[UserVersion] = None
+    _USER_VERSION = UserVersion(0, 4)  # The current / newest user version
+
+    def __init__(self, db_path: str, connection_name: Optional[str]=None):
+        self._database = QSqlDatabase.database(connection_name, open=False)
+        if not self._database.isValid():
```

Comment:
I don't really understand what you're going for with this entire "if". IMHO, we should raise an exception if a database with the same on-disk path already exists (creating the same database file twice seems like a mistake).

Reply:
The way I understand it, `QSqlDatabase` stores named connections to databases (they could even be to the same database). Now that we want to allow for multiple database instances to exist, we can't use the default name for all of them or they would just overwrite each other. If you try to create a connection with a name that already exists in `QSqlDatabase` (not even sure if that would happen) then shouldn't we return the existing connection?

We could use `db_path` as both the database name **and** the connection name to avoid the weird case where you pass an existing connection name but you changed the db_path. We would still need this `if` though.

Reply:
Raising an exception if a connection with the same path is also possible if we use db_path as the connection name (because `QSqlDatabase.database` lets you query by connection name, not by database name).

Reply:
Ah, I wasn't aware of the distinction between database name and connection name. After reviewing the [Qt docs](https://doc.qt.io/qt-5/qsqldatabase.html#details), things seem a bit clearer to me.

I was about to suggest not using the name mapping at all, but given that the docs say:

> **Warning: It is highly recommended that you do not keep a copy of the `QSqlDatabase` around as a member of a class, as this will prevent the instance from being correctly cleaned up on shutdown. If you need to access an existing `QSqlDatabase`, it should be accessed with `database()`. If you chose to have a `QSqlDatabase` member variable, this needs to be deleted before the `QCoreApplication` instance is deleted, otherwise it may lead to undefined behavior.

I suppose we should use the existing registration mechanism indeed, and avoid saving direct references to those databases.

However, I suggest we try to keep a clear distinction between:

1) Opening and registering a new database (ideally raising an exception if the same name already exists)
2) Accessing an existing database by name, perhaps even by having a class method like `Database.get('history')` or so (which would again raise an exception if it doesn't exist)

That way, we keep the initialization explicit rather than implicit, we make sure that we don't accidentally initialize the same database twice (thus avoiding the chance of discrepancies you're mentioning below), and we make sure we don't accidentally try to access a database with a wrong name (instead of automatically creating it).

What do you think? @rcorre any thoughts from your side?

Reply:
I hadn't seen that warning. So you suggest we don't even store `_database`, and instead retrieve it from `QSqlDatabase` every time our `Database` instance needs to use it?

Reply:
Yeah, sounds like we should do that indeed. Probably makes sense to wrap it in a `_database` method instead, though.

Reply:
I'd say that unless we have a current use case for multiple connections to a single database, any api for constructing a connection should throw if it already exists (as such a situation would be exceptional within the context of Qutebrowser as it works today). That behavior could always be changed later if there was a legitmate use case for multiple connections.

Reply:
Actually, just re-read what @The-Compiler said, and having separate APIs for creation/connection seems like a good way to avoid unintentional behavior

Reply:
Agreed. I used `path` both as the connection and the database name to avoid weird combinations like "multiple connections to the same database" or "same connection name but pointing to a different database"; an exception is thrown if that `path` is already in use. If we ever need any of those cases we can implement them at that point.